models/kernels/Nonlinear_Reservoir.hpp

```diff
@@ -0,0 +1,153 @@
+               / (parameters_struct.maximum_storage_meters - activation_threshold_meters), b);
+    };
+
+    private:
+    double a;
+    double b;
+};
+
+ 
+//This class is for a nonlinear reservoir that has zero, one, or multiple outlets. The nonlinear reservoir has parameters of minimum
+//and maximum storage height in meters and a state variable of current storage height in meters. A vector will be created that
+//stores pointers to the reservoir outlet objects. This class will also sort multiple outlets from lowest to highest activation
+//thresholds in the vector. The response_storage_meters function takes an inflow and cycles through the outlets from lowest to highest
+//activation thresholds and calls the outlet's function to return the discharge velocity in meters per second through the outlet.
+//The reservoir's storage is updated from velocities of each outlet and the delta time for the given timestep.
+class Nonlinear_Reservoir
+{
+    public:
+
+    //Constructor for a reservoir with no outlets.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters)
+    {
+        parameters.minimum_storage_meters = minimum_storage_meters;
+        parameters.maximum_storage_meters = maximum_storage_meters;
+        state.current_storage_height_meters = current_storage_height_meters;
+    }
+
+    //Constructor for a reservoir with only one outlet.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters, double a, 
+    double b, double activation_threshold_meters) : Nonlinear_Reservoir(minimum_storage_meters, maximum_storage_meters, 	current_storage_height_meters)
+    {
+        //Ensure that the activation threshold is less than the maximum storage
+        //if (activation_threshold_meters > maximum_storage_meters)
+        //TODO: Return appropriate error
+
+        this->outlets.push_back(Reservoir_Outlet(a, b, activation_threshold_meters));
+    }
+
+    //Constructor for a reservoir with multiple outlets.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters, std::vector <Reservoir_Outlet> &outlets) : Nonlinear_Reservoir(minimum_storage_meters, maximum_storage_meters, current_storage_height_meters)
+    {
+
+        this->outlets = outlets;
+	  
+        //Ensure that reservoir outlet activation thresholds are sorted from least to greatest height
+        sort(this->outlets.begin(), this->outlets.end(), [](const Reservoir_Outlet &left, const Reservoir_Outlet &right)
+        {              
+            return left.activation_threshold_meters < right.activation_threshold_meters;
+        });
+
+        //Ensure that the activation threshold is less than the maximum storage
+        for (auto& outlet : this->outlets) //pointer to outlets
```

Comment:
Since this is sorted now, you can simply look at at the last outlet and see if it is greater than max.  If you want to know more than one, just loop backwards until you are less than max.

Reply:
Ok thanks. Will change this too. @jdmattern-noaa changing.